### ISSUE TITLE
Vagrant broken by rubygems 3.1.2 upgrade: uninitialized constant I18n::Config (NameError)

### ISSUE THREAD

**Issue by andersk** [NixOS/nixpkgs#76629]
**Describe the bug**
Vagrant no longer works; every operation fails with an error `uninitialized constant I18n::Config (NameError)`. I bisected this regression to commit 298010982b8986f6669e2b762634dac48253d7ba “ruby: rubygems 3.0.6 -> 3.1.2” (#76495).

**To Reproduce**
```console
$ nix-shell -p vagrant --run 'vagrant version'
/nix/store/x2xagsdlb556kg1jzx0qrbc1wyjf8092-ruby2.6.5-json-2.1.0/lib/ruby/gems/2.6.0/gems/json-2.1.0/lib/json/common.rb:100: warning: already initialized constant JSON::NaN
/nix/store/p4hrycs0zaa9x0gsqylbk577ppnryixr-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/json-2.1.0/lib/json/common.rb:100: warning: previous definition of NaN was here
/nix/store/x2xagsdlb556kg1jzx0qrbc1wyjf8092-ruby2.6.5-json-2.1.0/lib/ruby/gems/2.6.0/gems/json-2.1.0/lib/json/common.rb:102: warning: already initialized constant JSON::Infinity
/nix/store/p4hrycs0zaa9x0gsqylbk577ppnryixr-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/json-2.1.0/lib/json/common.rb:102: warning: previous definition of Infinity was here
/nix/store/x2xagsdlb556kg1jzx0qrbc1wyjf8092-ruby2.6.5-json-2.1.0/lib/ruby/gems/2.6.0/gems/json-2.1.0/lib/json/common.rb:104: warning: already initialized constant JSON::MinusInfinity
/nix/store/p4hrycs0zaa9x0gsqylbk577ppnryixr-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/json-2.1.0/lib/json/common.rb:104: warning: previous definition of MinusInfinity was here
/nix/store/x2xagsdlb556kg1jzx0qrbc1wyjf8092-ruby2.6.5-json-2.1.0/lib/ruby/gems/2.6.0/gems/json-2.1.0/lib/json/common.rb:129: warning: already initialized constant JSON::UnparserError
/nix/store/p4hrycs0zaa9x0gsqylbk577ppnryixr-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/json-2.1.0/lib/json/common.rb:129: warning: previous definition of UnparserError was here
Traceback (most recent call last):
	11: from /nix/store/5y31ch80fwnvbz2lpc7shn0jgnf2mql1-vagrant-2.2.6/bin/.vagrant-wrapped:18:in `<main>'
	10: from /nix/store/5y31ch80fwnvbz2lpc7shn0jgnf2mql1-vagrant-2.2.6/bin/.vagrant-wrapped:18:in `load'
	 9: from /nix/store/5y31ch80fwnvbz2lpc7shn0jgnf2mql1-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/vagrant-2.2.6/bin/vagrant:88:in `<top (required)>'
	 8: from /nix/store/7l24siwk9yw0znjpcapq0rvpdghxmwsc-ruby-2.6.5/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
	 7: from /nix/store/7l24siwk9yw0znjpcapq0rvpdghxmwsc-ruby-2.6.5/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
	 6: from /nix/store/5y31ch80fwnvbz2lpc7shn0jgnf2mql1-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/vagrant-2.2.6/lib/vagrant.rb:278:in `<top (required)>'
	 5: from /nix/store/p4hrycs0zaa9x0gsqylbk577ppnryixr-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/i18n-1.1.1/lib/i18n.rb:55:in `load_path'
	 4: from /nix/store/p4hrycs0zaa9x0gsqylbk577ppnryixr-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/i18n-1.1.1/lib/i18n.rb:42:in `config'
	 3: from /nix/store/7l24siwk9yw0znjpcapq0rvpdghxmwsc-ruby-2.6.5/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
	 2: from /nix/store/7l24siwk9yw0znjpcapq0rvpdghxmwsc-ruby-2.6.5/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
	 1: from /nix/store/p4hrycs0zaa9x0gsqylbk577ppnryixr-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/i18n-1.1.1/lib/i18n/config.rb:5:in `<top (required)>'
/nix/store/p4hrycs0zaa9x0gsqylbk577ppnryixr-vagrant-2.2.6/lib/ruby/gems/2.6.0/gems/i18n-1.1.1/lib/i18n/config.rb:6:in `<module:I18n>': uninitialized constant I18n::Config (NameError)
```

**Expected behavior**
```console
$ nix-shell -I nixpkgs=channel:nixos-19.09 -p vagrant --run 'vagrant version'
Installed Version: 2.2.6
Latest Version: 2.2.6
 
You're running an up-to-date version of Vagrant!
```

**Metadata**
```console
$ nix run nixpkgs.nix-info -c nix-info -m
nix: src/libexpr/attr-set.hh:54: void nix::Bindings::push_back(const nix::Attr&): Assertion `size_ < capacity_' failed.
Aborted (core dumped)

$ what
what: command not found

$ nix-shell -p nix-info --run 'nix-info -m'
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.6, NixOS, 20.03pre206632.b0bbacb5213 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.1`
 - channels(root): `"nixos-20.03pre206632.b0bbacb5213, nixos-19.09-19.09.1687.c5d5561f772"`
 - channels(anders): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: [vagrant]
# a list of nixos modules affected by the problem
module:
```

cc @aneeshusa

**Comment by andersk** [comment]
I note that the `vagrant --version` “smoke test” in `installCheckPhase` is not sufficient to catch this problem, although `vagrant version` is. Perhaps `HOME="$(mktemp -d)" vagrant init --output - > /dev/null` would be even better?

**Comment by Ma27** [comment]
Just took a look at this as well (please note that I'm not a ruby developer, so I may be missing something).

`rubygems` 3.1.x seems to require bundler now[1] and vendors bundler 2.1.2[2] in version 3.1.2, however ruby 2.6 vendors bundler 1.17, so there's probably some kind of mismatch when using rubygems 3.1.2.

Not sure, but would it make sense to add different rubygems versions for different ruby versions?

[1] https://github.com/rubygems/rubygems/releases/tag/v3.1.0
[2] https://github.com/rubygems/rubygems/releases/tag/v3.1.2

**Comment by andersk** [comment]
@Ma27 I wrote that fix and submitted it as #76757. I also don’t know if it’s the right fix, but it does fix Vagrant.
